### PR TITLE
[cxxmodule] Don't declare fwd decls from rootmaps with modules

### PR DESCRIPTION
**Pull request by yamaguchi1024**
Fwd decls from rootmap files were injected to ROOT at the startup time.
However with modules we don't need any foward decls as we're preloading
all modules.

Our future plan is to remove rootmap files as a whole, but we still need
to support more features (see 2380). So for now we can remove what we
are not using.

*edit This patch also contains adding fHasCxxModule as a private member

This improves startup memory by 10 mbytes.

### CODE REVIEW

**Comment by yamaguchi1024** [comment]
@phsft-bot build just on slc6/gcc62 with flags -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

**Comment by yamaguchi1024** [comment]
We should not read decls from rootmap as they are already in pcm, no?

**Comment by vgvassilev** [comment]
The current patch essentially disabled rootmap files if we built with -Druntime_cxxmodules=On. That would break experiments which will still use the incremental approach where ROOT will rely on pcm files and their libraries in rootmaps. We should not read the rootmap files which have a preloaded module.

**Comment by pcanal** [comment]
Doesn't the precompiled module have to be generated by rootcling/genreflex to be useful for us?  Consequently shouldn't rootcling either generated a full rootmap **or** a short-rootmap and a precompiled module?

**Comment by yamaguchi1024** [comment]
I think Vassil's point is that there is still a little bit of un-modularized part in ROOT which is relying on rootmap files, and we should skip rootmap files only when we know that we have pcm. right?

**Comment by pcanal** [comment]
>  and we should skip rootmap files only when we know that we have pcm. right?

That's fine ... however I thought 'some' part of the rootmap file was still needed ...

**Comment by yamaguchi1024** [comment]
I'm just skipping forward decls from rootmaps, not all of them.

**Comment by yamaguchi1024** [comment]
@phsft-bot build just on slc6/gcc62 with flags -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

**Comment by yamaguchi1024** [comment]
@phsft-bot build just on slc6/gcc62 with flags -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

**Comment by yamaguchi1024** [comment]
@phsft-bot build!

File: core/metacling/src/TCling.h [hunk 1]
```diff
@@ -111,6 +111,8 @@ class TCling : public TInterpreter {
    TObjArray*      fRootmapFiles;     // Loaded rootmap files.
    Bool_t          fLockProcessLine;  // True if ProcessLine should lock gInterpreterMutex.
    Bool_t          fAllowLibLoad;     // True if library load is allowed (i.e. not in rootcling)
+   Bool_t          fHasCxxModule;     // True if C++ modules was enabled
```
Comment (vgvassilev) [inline core/metacling/src/TCling.h]: Please move this in a separate commit.

**Comment by yamaguchi1024** [comment]
@phsft-bot build just on slc6/gcc62 with flags -Druntime_cxxmodules=On -Dctest_test_exclude_none=on

File: core/metacling/src/TCling.cxx [hunk 2]
```diff
@@ -4964,6 +4964,26 @@ const char* TCling::TypeName(const char* typeDesc)
...
+   std::string rf = std::string(rootmapfile);
+
+   size_t s = rf.rfind("lib");
+   if (s != string::npos)
+      rf = rf.substr(s + 3);
+
+   s = rf.find(".rootmap");
+   if (s != string::npos)
+      rf.erase(s, 8);
+
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: We can replace this by:
```cpp
  llvm::StringRef moduleName = llvm::sys::path::filename(rootmapfile);
  moduleName.consume_front("lib");
  moduleName.consume_back("rootmap");
```

File: core/metacling/src/TCling.cxx [hunk 3]
```diff
@@ -4964,6 +4964,26 @@ const char* TCling::TypeName(const char* typeDesc)
...
+   Module *M = interp->getCI()->getPreprocessor().getHeaderSearchInfo().lookupModule(rf);
+   bool wasLoaded = false;
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: `bool wasLoaded = M && interp->getSema().isModuleVisible(M);`

File: core/metacling/src/TCling.cxx [hunk 4]
```diff
@@ -4964,6 +4964,26 @@ const char* TCling::TypeName(const char* typeDesc)
    return t;
 }
 
+static bool rMapHasPCM(const char* rootmapfile, cling::Interpreter* interp)
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: Please rename to `requiresRootMap` and negate the return result. Also, we should have a fast path for the non-modules case. Eg:
```cpp
  if (!fCxxModules)
    return true;
```

File: core/metacling/src/TCling.cxx [hunk 5]
```diff
@@ -4974,6 +4994,7 @@ int TCling::ReadRootmapFile(const char *rootmapfile, TUniqueString *uniqueString
 {
    // For "class ", "namespace ", "typedef ", "header ", "enum ", "var " respectively
    const std::map<char, unsigned int> keyLenMap = {{'c',6},{'n',10},{'t',8},{'h',7},{'e',5},{'v',4}};
+   bool hasPCM = rMapHasPCM(rootmapfile, fInterpreter);
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: Remove, we only use it once.

File: core/metacling/src/TCling.cxx [hunk 6]
```diff
@@ -4974,6 +4994,7 @@ int TCling::ReadRootmapFile(const char *rootmapfile, TUniqueString *uniqueString
...
 
    if (rootmapfile && *rootmapfile) {
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: Why don't we extend the condition here and make it `if (rootmapfile && *rootmapfile && requiresRootMap(...))`.

File: core/metacling/src/TCling.cxx [hunk 7]
```diff
@@ -5165,6 +5186,9 @@ Int_t TCling::LoadLibraryMap(const char* rootmapfile)
       fRootmapFiles->SetOwner();
       InitRootmapFile(".rootmap");
    }
+   bool hasPCM = false;
+   if (rootmapfile && *rootmapfile)
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: Likewise.

File: core/metacling/src/TCling.cxx [hunk 8]
```diff
@@ -5244,7 +5273,11 @@ Int_t TCling::LoadLibraryMap(const char* rootmapfile)
       }
    }
    if (rootmapfile && *rootmapfile) {
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: Likewise.

File: core/metacling/src/TCling.cxx (hunk 6 above)
Comment (yamaguchi1024) [reply core/metacling/src/TCling.cxx]: We just want to remove decls here, not everything in rootmap. So we cannot branch here.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
@phsft-bot build just on fedora28/native with flags -DCMAKE_BUILD_TYPE=Debug -DLLVM_BUILD_TYPE=Debug

**Comment by vgvassilev** [comment]
@phsft-bot build just on fedora28/native with flags -DCMAKE_BUILD_TYPE=Release -DLLVM_BUILD_TYPE=Release

**Comment by vgvassilev** [comment]
@phsft-bot build just on fedora28/native with flags -DCMAKE_BUILD_TYPE=Debug -DLLVM_BUILD_TYPE=RelWithDebInfo

**Comment by vgvassilev** [comment]
@phsft-bot build just on fedora28/native with flags -DCMAKE_BUILD_TYPE=Debug -DLLVM_BUILD_TYPE=RelWithDebInfo -Dccache=Off

**Comment by vgvassilev** [comment]
@phsft-bot build just on fedora28/native with flags -DCMAKE_BUILD_TYPE=Debug -DLLVM_BUILD_TYPE=Release -Dccache=ON

**Comment by vgvassilev** [comment]
@phsft-bot build just on fedora28/native with flags -DCMAKE_BUILD_TYPE=Debug -DLLVM_BUILD_TYPE=Release -Dccache=ON

**Comment by vgvassilev** [comment]
@phsft-bot build!
